### ISSUE TITLE
Create a Unit Testing project to speed up testing any code changes which can affect results

### ISSUE THREAD

**Issue by PCAssistSoftware** [tjscience/RoboSharp#156]
Whenever any changes are made to code involving areas such as the Progress Estimator (or Results) then usually myself or @RFBomb have to spend quite a long time manually running tests over and over again to ensure that the changes a) worked correctly and b) didn't have any adverse affect on the results shown by Progress Estimator or Results - this can be quite time consuming as it involves so many different factors.

As suggested by @RFBomb in a previous discussion it is proposed that at some point (time allowing) we create a unit testing project as below:-

---

In the unit test solution have a folder tree with a bunch of small files (0-2kb).

Then have the unit test generate and run commands, then compare the results of ProgressEstimator to actual results.

Static methods for:

- Generating the base command with source/destination
- Method to run the test in a normalized way and compare the results, handing them back to the unit test being run
- Method to purge the test destination

Then each UNIT TEST will simple call those static methods, turning options on/off as needed for the test.
 
This way we could test multiple job settings e.g. mirror, file / directory exclusions, attrib exclusions, min/max age, min/max size etc - all things which affect the results.

We can also test using different combinations of source files / folder names e.g. names containing spaces, special characters, zero byte files, all of which can affect results in some cases

---

Neither of us have ever built a unit testing service and it could be fairly complex to do so, but it would be quite an interesting area to learn more about, and would certainly save time in the long run testing any code changes.

There is no time frame for this - I am just adding it here so we can continue to discuss / plan it as and when time allows.

**Comment by RFBomb** [comment]
I whipped up a base to start from in VSCode (so I know theres issues with it, beacuse I was going by memory instead of working with the library, so no intellisense was aavilable to correct anything I typed up)

But if you read through the code, it should be fairly easy to understand the idea behind it. 

Basically, I created 2 new classes, one to act as the expected results, and one to represent a unit test's results. Then a method to compare IResults objects was created, and a method to compare and log IStatistics was created. A method to create a starter command and expected results object was also created, as well as a method to run the test. 

So each unit test should just call the method to generate the objects, modify them as needed, run the test, then evaluate the results object. The unit tests themselves need to be built, and these several classes should probably be divided up to their own files within the unit test project, but this should get the idea for the core down.

[UnitTest Mockup.zip](https://github.com/tjscience/RoboSharp/files/8191447/UnitTest.Mockup.zip)
